test/-ext-/string/test_rb_str_dup.rb

```diff
@@ -0,0 +1,16 @@
+require 'test/unit'
+require '-test-/string'
+
+class Test_RbStrDup < Test::Unit::TestCase
+  def test_nested_shared_non_frozen
+    str = Bug::String.rb_str_dup(Bug::String.rb_str_dup("a" * 50))
+    assert Bug::String.shared_string?(str)
+    refute Bug::String.sharing_with_shared?(str), '[Bug #15792]'
```

Comment:
Sorry to be late.
Could you use `assert_send [Bug::String, :shared_string?, str]` and `assert_not_send` for better failure messages?